pallets/omnipool/src/tests/buy.rs

```diff
@@ -394,6 +394,95 @@ fn simple_buy_with_fee_works() {
 		});
 }
 
+#[test]
+fn buy_should_emit_event_with_correct_fees() {
+	ExtBuilder::default()
+		.with_endowed_accounts(vec![
+			(Omnipool::protocol_account(), DAI, 1000 * ONE),
+			(Omnipool::protocol_account(), HDX, NATIVE_AMOUNT),
+			(LP2, 100, 2000 * ONE),
+			(LP3, 200, 2000 * ONE),
+			(LP1, 100, 1000 * ONE),
+		])
+		.with_registered_asset(100)
+		.with_registered_asset(200)
+		.with_asset_fee(Permill::from_percent(10))
+		.with_initial_pool(FixedU128::from(1), FixedU128::from(1))
+		.with_token(100, FixedU128::from(1), LP2, 2000 * ONE)
+		.with_token(200, FixedU128::from(1), LP3, 2000 * ONE)
+		.build()
+		.execute_with(|| {
+			let buy_amount = 50 * ONE;
+			let max_limit = 100 * ONE;
+			let expected_sold_amount = 58_823_529_411_766;
+
+			assert_ok!(Omnipool::buy(
+				RuntimeOrigin::signed(LP1),
+				200,
+				100,
+				buy_amount,
+				max_limit
+			));
+
+			expect_events(vec![Event::BuyExecuted {
+				who: LP1,
+				asset_in: 100,
+				asset_out: 200,
+				amount_in: expected_sold_amount,
+				amount_out: buy_amount,
+				hub_amount_in: 57142857142858,
+				hub_amount_out: 57142857142858,
+				asset_fee_amount: 5_555_555_555_556,
+				protocol_fee_amount: 0,
```

Comment:
you could maybe write a property test that checks that hub_amount_in - hub_amount_out = protocol_fee_amount. Not sure at what level the test should be written, but it would be good to have somewhere.

Reply:
(for both buys and sells)

Reply:
done as debug asserts in the code, so it runs on each invariant test run from now on.